master/buildbot/steps/source/base.py

```diff
@@ -279,22 +280,20 @@ def start(self):
                 # root is optional.
                 patch = s.patch
                 if patch:
-                    self.addCompleteLog("patch", bytes2unicode(patch[1]))
+                    yield self.addCompleteLog("patch", bytes2unicode(patch[1]))
             else:
                 log.msg("No sourcestamp found in build for codebase '{}'".format(self.codebase))
-                self.step_status.setText(
-                    ["Codebase", '{}'.format(self.codebase), "not", "in", "build"])
-                self.addCompleteLog("log",
-                                    "No sourcestamp found in build for codebase '{}'".format(
-                                            self.codebase))
+                self.descriptionDone = ["Codebase", '{}'.format(self.codebase),
+                                        "not", "in", "build"]
+                yield self.addCompleteLog("log",
+                                          "No sourcestamp found in build for codebase '{}'".format(
+                                               self.codebase))
                 return FAILURE
 
         else:
             revision = None
             branch = self.branch
             patch = None
 
-        d = self.startVC(branch, revision, patch)
-        d.addCallback(self.finished)
-        d.addErrback(self.failed)
-        return None
+        res = yield self.run_vc(branch, revision, patch)
```

Comment:
this means that any custom VC will stop working. is this what we want?

Reply:
Hmm. So the problem is that the base class is being changed to new style step and any sub-classes must be changed in lock-step. Just calling startVC would just hide bugs.

The same problem is present in other steps. The source steps are probably unique in that the base class packs a lot of functionality.

I wonder if we could work around this by implementing a base run_vc that would check for existence of startVC and if such exists, would setup the step for the old style much like BuildStep.run does. This temporary solution would be removed on v3.0

Reply:
I would rather avoid another pile of hacks like I did in the base step. As the migration is pretty straighforward (just changing method name and yielding addLogForRemoteCommands, maybe we should just warn and create a small migration doc.
I would not expect so many complicated source steps anyway.

Reply:
We also need to return the result as a return value from `startVC` instead of calling `finished` or `failed`. This makes it much more difficult to support the old source steps.

I think the best would be to check whether `startVC` exists and throw an exception without even trying to run the step.

We can recommend to include the source step into the buildbot proper in the exception message :-)

Reply:
@tardyp: Fixed in 7570d9c.